service/history/workflow/update/update.go

```diff
@@ -156,34 +160,96 @@ func newCompleted(
 	return upd
 }
 
+// WaitLifecycleStage waits until the Update has reached at least `waitStage` or
+// a timeout. If the Update reaches `waitStage` with no timeout, the most
+// advanced stage known to have been reached is returned, along with the outcome
+// if any. If there is a timeout due to the supplied soft timeout, then
+// unspecified stage and nil outcome are returned, without an error. If there is
+// a timeout due to context deadline expiry, then the error is returned as usual.
+func (u *Update) WaitLifecycleStage(
+	ctx context.Context,
+	waitStage enumspb.UpdateWorkflowExecutionLifecycleStage,
+	softTimeout time.Duration) (stage enumspb.UpdateWorkflowExecutionLifecycleStage, outcome *updatepb.Outcome, err error) {
+
+	stage = enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED
+	switch waitStage {
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED:
+		return u.waitLifecycleStage(ctx, u.WaitAccepted, softTimeout)
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED:
+		return u.waitLifecycleStage(ctx, u.WaitOutcome, softTimeout)
+	default:
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, serviceerror.NewUnimplemented(
+			fmt.Sprintf("%v is not implemented", waitStage))
+	}
+}
+
+func (u *Update) waitLifecycleStage(
+	ctx context.Context,
+	waitFn func(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error),
+	softTimeout time.Duration) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
+
+	type result struct {
+		stage   enumspb.UpdateWorkflowExecutionLifecycleStage
+		outcome *updatepb.Outcome
+		err     error
+	}
+	ch := make(chan result, 1)
+
+	go func() {
```

Comment:
I thought about this some more and you should be able to avoid this extra coroutine with something like the following:

```go
innerCtx, cancel := context.WithTimeout(ctx, softTimeout)
defer cancel()
stage, outcome, err := waitFn(innerCtx)
if ctx.Err() != nil {
  // Handle a context deadline expiry as usual.
  return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, ctx.Err()
}
if innerCtx.Err() != nil {
  return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, nil
}
return stage, outcome, err
```

Reply:
Ah, of course, that's much nicer, thanks! I'll be more familiar with Go context's after this PR. (The tests pass before and after this change btw, so we have good evidence that this is a purely synonymous refactor).